File: backend/app/schemas/golden_template.py
"""Pydantic schemas for Fase 17 — Golden Config Templates."""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class TemplateVariable(BaseModel):
    key: str
    type: str  # ip, cidr, string, integer, hostname
    label: str
    required: bool = True
    default: str | None = None
    hint: str | None = None


class GoldenTemplateCreate(BaseModel):
    name: str
    description: str | None = None
    vendor: str = "any"
    category: str
    variables: list[TemplateVariable] = []
    content: str = ""
    change_note: str | None = None


class GoldenTemplateUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    vendor: str | None = None
    category: str | None = None
    variables: list[TemplateVariable] | None = None
    content: str | None = None
    change_note: str | None = None


class GoldenTemplateVersionRead(BaseModel):
    id: UUID
    version: int
    change_note: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class GoldenTemplateRead(BaseModel):
    id: UUID
    tenant_id: UUID | None
    name: str
    description: str | None
    vendor: str
    category: str
    variables: list[dict]
    content: str
    version: int
    is_active: bool
    is_system: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class GoldenTemplateSummary(BaseModel):
    id: str
    tenant_id: str | None
    name: str
    description: str | None
    vendor: str
    category: str
    variable_count: int
    version: int
    is_system: bool
    created_at: str
    updated_at: str


class RenderRequest(BaseModel):
    variable_values: dict[str, str] = {}


class RenderResponse(BaseModel):
    content: str
    unresolved: list[str]


class ApplyRequest(BaseModel):
    device_id: str
    variable_values: dict[str, str] = {}


class ApplyResponse(BaseModel):
    status: str  # "applied" | "manual" | "error"
    message: str
    output: str | None = None
    commands: str | None = None  # returned when status="manual"


class DivergenceRequest(BaseModel):
    device_id: str
    variable_values: dict[str, str] = {}


class DivergenceItem(BaseModel):
    section: str
    value: str
    status: str  # "missing" | "extra"


class DivergenceResponse(BaseModel):
    device_id: str
    template_id: str
    vendor: str
    items: list[DivergenceItem]
    summary: dict[str, int]
    rendered_preview: str
    supported: bool
    message: str | None = None


class PrefillResponse(BaseModel):
    variable_values: dict[str, str]

File: frontend/src/types/goldenTemplate.ts
export interface TemplateVariable {
  key: string;
  type: "ip" | "cidr" | "string" | "integer" | "hostname";
  label: string;
  required: boolean;
  default?: string;
  hint?: string;
}

export interface GoldenTemplateSummary {
  id: string;
  tenant_id: string | null;
  name: string;
  description: string | null;
  vendor: string;
  category: string;
  variable_count: number;
  version: number;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface GoldenTemplateRead {
  id: string;
  tenant_id: string | null;
  name: string;
  description: string | null;
  vendor: string;
  category: string;
  variables: TemplateVariable[];
  content: string;
  version: number;
  is_active: boolean;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface GoldenTemplateVersionRead {
  id: string;
  version: number;
  change_note: string | null;
  created_at: string;
}

export interface GoldenTemplateCreate {
  name: string;
  description?: string;
  vendor: string;
  category: string;
  variables: TemplateVariable[];
  content: string;
  change_note?: string;
}

export interface GoldenTemplateUpdate {
  name?: string;
  description?: string;
  vendor?: string;
  category?: string;
  variables?: TemplateVariable[];
  content?: string;
  change_note?: string;
}

export interface RenderResponse {
  content: string;
  unresolved: string[];
}

export interface ApplyResponse {
  status: "applied" | "manual" | "error";
  message: string;
  output?: string;
  commands?: string;
}

export interface DivergenceItem {
  section: string;
  value: string;
  status: "missing" | "extra";
}

export interface DivergenceResponse {
  device_id: string;
  template_id: string;
  vendor: string;
  items: DivergenceItem[];
  summary: { missing: number; extra: number; total: number };
  rendered_preview: string;
  supported: boolean;
  message?: string;
}
